### PR TITLE
Export applications with HESA data

### PR DESCRIPTION
**Pull request by steventux**
## Context

_HEIs need to report information about new trainees to HESA, who then send this data back to DfE for statistical purposes. To reduce the burden on providers needing to report back to the department, we can present the relevant info in a friendly format._

## Changes proposed in this pull request

Adds a default applications export in CSV format. This exports application data in the following format which includes HESA specific codes:

id | status | first name | last name | date of birth | nationality | domicile | email address | recruitment cycle | provider code | accredited body | course code | site code | study mode | SBJCA | QLAIM | FIRSTDEG | DEGTYPE | DEGSBJ | DEGCLSS | institution country | DEGSTDT | DEGENDDT | institution details | sex | disabilities | ethnicity
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
JV5729 | pending_conditions | Shalon | Kovacek | 1968-08-16 | British | GB | shalon.kovacek@example.com | 2020 | 24J |   | AT75 | B | 2 | 100511 | 21 | 1 | 209 | 100321 | 98 | GB | 2020 | 2013 | 182 | not specified | not specified | 22
AH2247 | offer | Jerome | Leuschke | 2002-06-07 | British | GB | jerome.leuschke@example.com | 2020 | 24J |   | 2KBS | B | 2 | 100511 | 21 | 1 | 209 | 100846 | 7 | GB | 2020 | 1980 | 62 | no data | no data | no data
RR4296 | offer_deferred | Mitchel | Erdman | 1982-12-06 | British | GB | mitchel.erdman@example.com | 2020 | 24J |   | 2N2M | B | 2 | 100511 | 21 | 1 | 87 | 100350 | 6 | GB | 2020 | 2014 | 92 | 1 | 53 55 | 80
QB1437 | offer | Chantelle | Wehner | 1981-02-23 | British | GB | chantelle.wehner@example.com | 2020 | 1JA | Keele and North Staffordshire Teacher Education | 2ZRM | - | 2 | 100511 | 21 | 1 | 86 | 100456 | 12 | GB | 2020 | 1995 | 95 | not specified | 58 96 | 10
HD4611 | recruited | Jeraldine | Bode | 1967-08-19 | British | GB | jeraldine.bode@example.com | 2020 | 1JA | Keele and North Staffordshire Teacher Education | 2ZRM | - | 2 | 100511 | 21 | 1 | 93 | 100166 | 13 | GB | 2020 | 2010 | 22 | no data | no data | no data
HR6343 | pending_conditions | Larry | Gerlach | 1977-07-15 | British | GB | larry.gerlach@example.com | 2020 | 1JA | Keele and North Staffordshire Teacher Education | 2ZRN | U | 2 | 100511 | 21 | 1 | 93 | 100846 | 14 | GB | 2020 | 2021 | 44 | not specified | not specified | 80

image

## Guidance to review

- I think we should also do the same as this PR https://github.com/DFE-Digital/apply-for-teacher-training/pull/3147 and background the export generation
- Are there further permissions checks necessary for this feature? Currently a provider user can export applications for all their providers.
- The route/controller naming could be better but I couldn't think of a better alternative. 

## Link to Trello card

https://trello.com/c/BS5j6MeN/2795-%F0%9F%8F%88-let-users-export-hesa-specific-application-data

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training#azure-hosting-devops-pipeline)

### CODE REVIEW

File: app/services/provider_interface/hesa_data_export.rb [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+      'conditions_met' => [:recruited],
+      'deferred' => [:offer_deferred],
+      'declined' => [:offer_declined],
+      'rejected' => [:offer_rejected],
+      'rejected_automatically' => [],
+      'application_withdrawn' => [:withdrawn],
+      'offer_withdrawn' => [:offer_withdrawn],
+    }.freeze
+
+    def initialize(export_options: {})
```
Comment (duncanjbrown) [inline app/services/provider_interface/hesa_data_export.rb:18]: It looks like we only ever use `provider_ids` from this hash? Would it be better to simply have these as named params as it's extra ceremony for the caller, eg https://github.com/DFE-Digital/apply-for-teacher-training/pull/3135/files#diff-bdff2c50fb9f23305fcf23f9de8fd2382f263023b7abf9be1b7ecd4ec7d0cd5dR8

**Review by duncanjbrown — COMMENTED** [review]
Interface question inline! Going to run locally and have a look

File: app/services/provider_interface/hesa_data_export.rb (hunk 1 above)
Comment (steventux) [reply app/services/provider_interface/hesa_data_export.rb:18]: Yes for sure, this is a relic of attempting the full-fat version with status and recruitment years filters, we decided not to include these initially. I can strip this back.

Comment (steventux) [reply app/services/provider_interface/hesa_data_export.rb]: @duncanjbrown I've simplified this to only pass `provider_ids` and have removed the redundant form class.

File: app/services/provider_interface/hesa_data_export.rb [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+      ]
+    end
+
+    def csv_headings
+      [
+        'id', 'status', 'first name', 'last name', 'date of birth', 'nationality', 'domicile', 'email address',
+        'recruitment cycle', 'provider code', 'accredited body', 'course code', 'site code', 'study mode', 'SBJCA',
+        'QLAIM', 'FIRSTDEG', 'DEGTYPE', 'DEGSBJ', 'DEGCLSS', 'institution country', 'DEGSTDT', 'DEGENDDT',
+        'institution details', 'sex', 'disabilities', 'ethnicity'
+      ]
```
Comment (tijmenb) [inline app/services/provider_interface/hesa_data_export.rb:97]: This is not unlikely to go out of sync with the data array. Can we generate a hash instead like some of the other exports?

https://github.com/DFE-Digital/apply-for-teacher-training/blob/da381583a804669e7651849cfe4d54d3988af557/app/services/support_interface/application_choices_export.rb#L6-L23

You can use it with SafeCSV like this:

https://github.com/DFE-Digital/apply-for-teacher-training/blob/168da216862bfdb4a588d683d2358cb77a5a2a2c/app/workers/data_exporter.rb#L19-L22

Comment (steventux) [reply app/services/provider_interface/hesa_data_export.rb:96]: @tijmenb :+1: https://github.com/DFE-Digital/apply-for-teacher-training/pull/3135/commits/88021a8b53b528ac2655785432a5852faa28c012

File: app/services/provider_interface/hesa_data_export.rb [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+      applications = ApplicationChoice
+        .includes(
+          :course,
+          :provider,
+          :site,
+          course_option: { course: %i[provider accredited_provider] },
+          application_form: %i[candidate application_qualifications],
+        ).where(
+          'candidates.hide_in_reporting' => false,
+          'status' => %w[accepted offer pending_conditions],
```
Comment (duncanjbrown) [inline app/services/provider_interface/hesa_data_export.rb:17]: Do we want to include `offer` here — maybe use `ApplicationStateChange::ACCEPTED_STATES`?

Comment (steventux) [reply app/services/provider_interface/hesa_data_export.rb]: :+1: @duncanjbrown https://github.com/DFE-Digital/apply-for-teacher-training/pull/3135/commits/14cb7d102cfe29664eaee6492f2bb8c14e6e5b1a

**Comment by duncanjbrown** [comment]
I just tried to run locally on freshly-generated test data and got this error:

Screenshot 2020-10-21 at 11 07 52

**Comment by duncanjbrown** [comment]
> I just tried to run locally on freshly-generated test data and got this error:
> 
> Screenshot 2020-10-21 at 11 07 52

I wonder if this is test data at fault, though?

**Comment by steventux** [comment]
@duncanjbrown it seems like it _should_ be array data https://github.com/DFE-Digital/apply-for-teacher-training/search?p=1&q=hesa_disabilities but we could be defensive about this.

**Review by duncanjbrown — APPROVED** [review]
LGTM!

**Pull request merged**